fw/websocket.c

```diff
@@ -0,0 +1,263 @@
+/**
+ *		Tempesta FW
+ *
+ * Websocket proxy protocol implementation for Tempesta FW.
+ *
+ * Copyright (C) 2015-2022 Tempesta Technologies, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+#undef DEBUG
+#if DBG_WS > 0
+#define DEBUG DBG_WS
+#endif
+
+#include "cfg.h"
+#include "connection.h"
+#include "websocket.h"
+#include "http.h"
+#include "http_msg.h"
+#include "sync_socket.h"
+#include "tempesta_fw.h"
+
+#define TFW_CONN_HTTP_TYPE(c)	\
+	(TFW_FSM_TYPE(TFW_CONN_TYPE(c)) & (TFW_FSM_HTTP | TFW_FSM_HTTPS))
+
+/**
+ * Global level websocket configuration.
+ *
+ * @client_ws_timeout	- timeout between two consecutive client sends before
+ * 			  connection close;
+ */
+static struct tfw_cfg_ws_t {
+	int client_ws_timeout;
+} tfw_cfg_ws;
+
+typedef struct tfw_cfg_ws_t TfwCfgWs;
+
+/**
+ * Process data for websocket connection without any introspection and
+ * analisis of the protocol. Just send it as is.
+ *
+ * Create light http msg with Conn_Srv flag just to append skb to it and forward
+ * it further. Client messages contain many fields not needed for websocket
+ * protocol.
+ */
+int
+tfw_ws_msg_process(TfwConn *conn, struct sk_buff *skb)
+{
+	int r;
+	TfwHttpMsg *hm = __tfw_http_msg_alloc(Conn_Srv, false);
+	if (!hm) {
+		__kfree_skb(skb);
+		return T_DROP;
+	}
+	ss_skb_queue_tail(&hm->msg.skb_head, skb);
+
+	if (TFW_CONN_TYPE(conn) & Conn_Srv) {
+		if ((r = tfw_cli_conn_send((TfwCliConn *)conn->pair,
+					   (TfwMsg *)hm)))
+			T_DBG("%s: cannot send data to client via websocket\n",
+			      __func__);
+	} else {
+		if ((r = tfw_connection_send(conn->pair, (TfwMsg *)hm)))
+			T_DBG("%s: cannot send data to backend via websocket\n",
+			      __func__);
+	}
+
+	if (r) {
+		tfw_connection_close(conn, true);
+	}
+
+	tfw_http_msg_free(hm);
```

Comment:
I think we don't need the dummy `hm` and call `tfw_connection_send()`, which just wraps our `tfw_ws_conn_send()`, which is in turn wraps `ss_send()`, and just can proxy the `skb` with `ss_send()` on the paired connection.

Reply:
We need dummy http message at least for secure websockets case to properly call `tfw_tls_conn_send()`.

All websocket hooks downcall to proper http or https hooks.